### PR TITLE
mpMetrics-3.0 NO SHIP code drop (JAX-RS relevant changes in two other separate PRs  )

### PR DESCRIPTION
**Pull request by Channyboy**
MicroProfile Metrics features:
auto - com.ibm.websphere.appserver.mpMetrics-3.0-monitor-1.0
public - com.ibm.websphere.appserver.mpMetrics-3.0
private - com.ibm.websphere.appserver.org.eclipse.microprofile.metrics-3.0

- Using a snapshot of the MicroProfile Metric API has it has not been released yet or RCed. Cannot retrieve from maven. (i.e the src folder of `com.ibm.websphere.org.eclipse.microprofile.metrics.3.0`)

- Introduced Stereotype support (MetricResolver#hasMetricAnnotation, MetricResolver#getAnnotation)

- New MetricRegistry API. It is now an interface and new methods are added.

- Concept of Re-usability removed (MetricResolver class modified - remove use of getReusable method)

- SimpleTimer now tracks high and lows from the "previous" full/complete minute. "Previous" is defined as the last time the simple timer was queried which could be a few minutes ago. This is similar to the behaviour of the "previous" minute tracking of the ConcurrentGauge metric.

#build

### CODE REVIEW

File: dev/com.ibm.ws.microprofile.metrics.cdi.3.0/src/io/astefanutti/metrics/cdi30/MetricResolver.java [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+        String[] tags = of.tags();
+        MetricID metricID = new MetricID(name, Utils.tagsToTags(tags));
+
+        Metadata existingMetadata = registry.getMetadata(name);
+
+        /*
+         * If the current metadata does not equal an existing metadata (of matching name).. - Then we can NOT re-use.
+         * Type check is conducted when registering/retrieving metric.
+         */
+        if ((existingMetadata != null && !existingMetadata.equals(of.metadata()))) {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.cdi.3.0/src/io/astefanutti/metrics/cdi30/MetricResolver.java:361]: Removal of reusability only checks if metadata is different.
There also use to exist code that retrieved the `reusable` value from the annotation. But that has been removed as is not apparant in this diff.

File: dev/com.ibm.ws.microprofile.metrics.cdi.3.0/src/io/astefanutti/metrics/cdi30/MetricResolver.java [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+        MetadataBuilder mdb = Metadata.builder().withName(name).withType(this.getType(annotation)).withUnit(this.getUnit(annotation)).withDescription(this.getDescription(annotation)).withDisplayName(this.getDisplayname(annotation));
+
+        String[] tags = this.getTags(annotation);
+
+        Of<T> of = new DoesHaveMetric<>(annotation, name, mdb.build(), initialDiscovery, tags);
+        checkReusable(of);
+        return of;
+    }
+
+    protected <T extends Annotation> boolean hasMetricAnnotationBean(Class<T> metric, Class<?> bean) {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.cdi.3.0/src/io/astefanutti/metrics/cdi30/MetricResolver.java:119]: this method added to support CDI stereotypes

File: dev/com.ibm.ws.microprofile.metrics.cdi.3.0/src/io/astefanutti/metrics/cdi30/MetricResolver.java [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+            Class<? extends Annotation> annotationType = annotation.annotationType();
+            if (beanManager.isStereotype(annotationType) && annotationType.isAnnotationPresent(metric)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    protected <T extends Annotation> T getAnnotationBean(Class<T> metric, Class<?> bean) {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.cdi.3.0/src/io/astefanutti/metrics/cdi30/MetricResolver.java:139]: this method added to support CDI stereotypes

File: dev/com.ibm.ws.microprofile.metrics.cdi.3.0/src/io/astefanutti/metrics/cdi30/MetricResolver.java [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+                }
+            }
+        }
+
+        return null;
+    }
+
+    protected <E extends Member & AnnotatedElement, T extends Annotation> Of<T> beanResolverOf(E element, Class<T> metric, Class<?> bean) {
+
+        if (hasMetricAnnotationBean(metric, bean)) {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.cdi.3.0/src/io/astefanutti/metrics/cdi30/MetricResolver.java:166]: 166-167 using the above two new methods.
It use to directly check the bean if the annotation was present and retrieve it. That of course did not support stereotypes.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/SimpleTimer30Impl.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+    @Override
+    public Duration getElapsedTime() {
+        maybeStartNewMinute();
+        return elapsedTime;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Duration getMaxTimeDuration() {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/SimpleTimer30Impl.java:166]: The implementation for calculating/retrieving the previous max high and low is similar to the implementation from ConcurrentGauge.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/Timer30Impl.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+import com.ibm.ws.microprofile.metrics.impl.HistogramImpl;
+import com.ibm.ws.microprofile.metrics.impl.MeterImpl;
+import com.ibm.ws.microprofile.metrics.impl.Reservoir;
+import com.ibm.ws.microprofile.metrics.impl.TimerImpl;
+
+/**
+ * A timer metric which aggregates timing durations and provides duration statistics, plus
+ * throughput statistics via {@link Meter}.
+ */
+public class Timer30Impl implements Timer {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/Timer30Impl.java:45]: Changes/Additions to the Timer to use the `Duration` object. (as per the spec)

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java [hunk 7]
```diff
@@ -0,0 +1,1090 @@
...
+        for (Entry<MetricID, Metric> entrySet : metricsMID.entrySet()) {
+            if (entrySet.getKey().getName().equals(name) && !metricClass.isAssignableFrom(entrySet.getValue().getClass())) {
+                throw new IllegalArgumentException(name + " is already used for a different type of metric");
+            }
+        }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends Metric> SortedMap<MetricID, T> getMetrics(Class<T> ofType, MetricFilter filter) {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java:754]: new method.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java [hunk 8]
```diff
@@ -0,0 +1,1090 @@
...
+        return (SortedMap<MetricID, T>) getMetrics(
+                                                   (metricID, metric) -> filter.matches(metricID, metric)
+                                                                         && ofType.isAssignableFrom(metric.getClass()));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SortedMap<MetricID, Metric> getMetrics(MetricFilter filter) {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java:762]: new method.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java [hunk 9]
```diff
@@ -0,0 +1,1090 @@
...
+        return Collections.unmodifiableMap(metricsMID);
+    }
+
+    @Override
+    public Map<String, Metadata> getMetadata() {
+        return Collections.unmodifiableMap(metadataMID);
+    }
+
+    @Override
+    public Metadata getMetadata(String name) {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java:784]: new method.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java [hunk 10]
```diff
@@ -0,0 +1,1090 @@
...
+        return this.simpleTimer(metadata, null);
+    }
+
+    @Override
+    public SimpleTimer simpleTimer(Metadata metadata, Tag... tags) {
+        return getOrAdd(metadata, MetricBuilder30.SIMPLE_TIMER, tags);
+    }
+
+    @Override
+    public SortedMap<MetricID, SimpleTimer> getSimpleTimers() {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java:1000]: All new methods below. More overloaded methods to register/retreive metrics. They end up calling the other existing overloaded methods.

**Comment by Channyboy** [comment]
Comments to help with review. 
Bulk of the change are the copying of necessary classes (i.e direct dependency to the metrics API and can't be extended) plus the numeral version changes as necessary.

However, the "meatier" changes that are the main focus of the developing MP metrics 3.0 spec are coming in the next 2 PRs which relate to JAXRS.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        results.put(Constants.PERCENTILE_999TH + tags, timer.getSnapshot().get999thPercentile() / conversionFactor);
+
+        return results;
+    }
+
+    /*
+     * To facilitate changes into the MicroProfile Specification in 3.0 we need to return a value that is an
+     * Object because the maxTimeDuration or minTimeDuration could be a NaN object.
+     */
+    public static Map<String, Object> getSimpleTimerNumbersAsObjects(SimpleTimer simpleTimer, String tags, double conversionFactor) {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java:138]: Needed a new method that returns a map with Object values. This is due to the possibility of a null value being returned.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/writer/JSONMetricWriter30.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+            }
+        }
+        return jsonObject;
+    }
+
+    /*
+     * This method compared to getJsonFromMap returns a map with a value as an Object.. This is because
+     * the value can be a NaN object.
+     */
+    protected JSONObject getJsonFromMapAsObject(Map<String, Object> metricMap, String metricName, JSONObject parentJSONObject) {
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/writer/JSONMetricWriter30.java:154]: Needed a new method that returns a map with Object values. This is due to the possibility of a null value being returned.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/writer/JSONMetricWriter30.java [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+            } else if (Gauge.class.isInstance(metric)) {
+                try {
+                    jsonObject.put(metricNameWithTags, ((Gauge) metric).getValue());
+                } catch (IllegalStateException e) {
+                    // The forwarding gauge is likely unloaded. A warning has already been emitted
+                }
+            } else if (Timer.class.isInstance(metric)) {
+                jsonObject.put(metricName, getJsonFromMap(Util30.getTimerNumbers((Timer) metric, tags, conversionFactor), metricName, jsonObject));
+            } else if (SimpleTimer.class.isInstance(metric)) {
+                jsonObject.put(metricName, getJsonFromMapAsObject(Util30.getSimpleTimerNumbersAsObjects((SimpleTimer) metric, tags, conversionFactor), metricName, jsonObject));
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/writer/JSONMetricWriter30.java:138]: Use of new methods where Object values are returned over Number values.

**Comment by Channyboy** [comment]
#bulid
#run-libby-bot

**Comment by Channyboy** [comment]
#build
#run-libby-bot

**Comment by Channyboy** [comment]
NLS props review not needed as these files are a copy of an earlier version (i.e mpMetrics-2.3 and occurrences of version references have been updated to 3.0)

**Comment by Channyboy** [comment]
The failures do not relate to this PR as this code is "disabled". The failures are pre-existing failures that exist in other Personal Builds and the Continuous build.

File: dev/com.ibm.ws.microprofile.metrics.3.0.monitor/bnd.bnd [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+javac.source: 1.8
+javac.target: 1.8
+
+Bundle-Name: com.ibm.ws.microprofile.metrics.3.0.monitor
+Bundle-SymbolicName: com.ibm.ws.microprofile.metrics.3.0.monitor
+Bundle-Description: MicroProfile Metrics Monitor Bridge Auto Feature, version ${bVersion}
+
+Import-Package: \
+	javax.enterprise.*;version="[1.1,3)",\
+	org.eclipse.microprofile.metrics; version="[2.3.0,4.0.0)",\
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0.monitor/bnd.bnd:23]: Update version to `[3.0.0,4.0.0)`

File: dev/com.ibm.ws.microprofile.metrics.3.0.monitor/bnd.bnd [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
+
+-dsannotations: com.ibm.ws.microprofile.metrics.monitor.MonitorMetricsHandler
+
+Service-Component: \
+
+-buildpath: \
+	com.ibm.ws.logging;version=latest,\
+	com.ibm.ws.microprofile.metrics.2.0;version=latest,\
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0.monitor/bnd.bnd:45]: Is 2.0 still needed?

File: dev/com.ibm.websphere.org.eclipse.microprofile.metrics.3.0/bnd.bnd [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+Import-Package: \
+  javax.enterprise.util; version="[1.1,3)",\
+  *
+
+Export-Package: \
+  org.eclipse.microprofile.metrics;version=3.0, \
+  org.eclipse.microprofile.metrics.annotation;version=3.0
+
+#Include-Resource: \
+#  @${repo;org.eclipse.microprofile.metrics:microprofile-metrics-api;2.3;EXACT}
```
Comment (fmhwong) [inline dev/com.ibm.websphere.org.eclipse.microprofile.metrics.3.0/bnd.bnd:28]: Change 2.3 to 3.0

File: dev/com.ibm.websphere.org.eclipse.microprofile.metrics.3.0/bnd.bnd [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+
+Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version>=1.8))"
+
+WS-TraceGroup: METRICS
+
+-buildpath: \
+	com.ibm.websphere.javaee.cdi.1.2;version=latest,\
+	com.ibm.websphere.javaee.interceptor.1.2;version=latest,\
+	com.ibm.ws.org.osgi.annotation.versioning;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.config.1.3;version=latest
```
Comment (fmhwong) [inline dev/com.ibm.websphere.org.eclipse.microprofile.metrics.3.0/bnd.bnd:38]: Should config be 1.4?

File: dev/com.ibm.websphere.org.eclipse.microprofile.metrics.3.0/bnd.bnd [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+#Include-Resource: \
+#  @${repo;org.eclipse.microprofile.metrics:microprofile-metrics-api;2.3;EXACT}
+
+Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version>=1.8))"
+
+WS-TraceGroup: METRICS
+
+-buildpath: \
+	com.ibm.websphere.javaee.cdi.1.2;version=latest,\
```
Comment (fmhwong) [inline dev/com.ibm.websphere.org.eclipse.microprofile.metrics.3.0/bnd.bnd:35]: Should cdi be 2.0?

File: dev/com.ibm.ws.microprofile.metrics.3.0.monitor/bnd.bnd [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+	com.ibm.ws.microprofile.metrics.3.0;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.metrics.3.0;version=latest,\
+	com.ibm.websphere.org.osgi.core,\
+    com.ibm.websphere.org.osgi.service.component,\
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0.monitor/bnd.bnd:49]: Fix uneven indentation

File: dev/com.ibm.ws.microprofile.metrics.3.0/bnd.bnd [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+-buildpath: \
+	com.ibm.websphere.org.osgi.core;version=latest,\
+	com.ibm.ws.container.service;version=latest,\
+	com.ibm.ws.adaptable.module;version=latest,\
+	com.ibm.ws.logging;version=latest,\
+	com.ibm.ws.classloading;version=latest,\
+	com.ibm.websphere.org.osgi.service.component;version=latest,\
+	com.ibm.wsspi.org.osgi.service.component.annotations;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.config.1.3;version=latest,\
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/bnd.bnd:64]: Should config use 1.4?

File: dev/com.ibm.ws.microprofile.metrics.3.0/bnd.bnd [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+	com.ibm.websphere.rest.handler;version=latest,\
+	com.ibm.ws.app.manager.wab;version=latest,\
+	com.ibm.ws.rest.handler;version=latest,\
+	com.ibm.ws.webcontainer;version=latest,\
+	com.ibm.ws.container.service;version=latest,\
+	com.ibm.ws.container.service.compat;version=latest,\
+	com.ibm.websphere.javaee.annotation.1.2;version=latest,\
+	com.ibm.ws.org.osgi.annotation.versioning;version=latest,\
+	com.ibm.websphere.javaee.interceptor.1.2;version=latest,\
+	com.ibm.websphere.javaee.cdi.1.2;version=latest, \
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/bnd.bnd:81]: Should cdi be 2.0?

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/BaseMetrics.java [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+            try {
+                if (isComposite) {
+                    CompositeData value = (CompositeData) mbs.getAttribute(new ObjectName(objectName), attribute);
+                    return (T) value.get(subAttribute);
+                } else {
+                    T value = (T) mbs.getAttribute(new ObjectName(objectName), attribute);
+                    return value;
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/BaseMetrics.java:189]: Use Tr to log an error

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/BaseMetrics.java [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+            this.attribute = attribute;
+        }
+
+        @Override
+        public long getCount() {
+            try {
+                Number value = (Number) mbs.getAttribute(new ObjectName(objectName), attribute);
+                return value.longValue();
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/BaseMetrics.java:209]: Use Tr to log an error

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+import com.ibm.ws.microprofile.metrics.exceptions.NoSuchRegistryException;
+import com.ibm.ws.microprofile.metrics23.helper.Util23;
+
+/**
+ *
+ */
+public class Util30 extends Util23 {
+    private static final TraceComponent tc = Tr.register(Util30.class);
+
+    public static Map<MetricID, Metric> getMetricsAsMap(String registryName, String metricName) throws NoSuchRegistryException, NoSuchMetricException, EmptyRegistryException {
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java:40]: Do you expect FFDCs for the thrown exceptions?

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+            throw new NoSuchMetricException();
+        } else {
+            for (MetricID tmid : potentialMatches) {
+                returnMap.put(tmid, registry.getMetric(tmid));
+            }
+        }
+        return returnMap;
+    }
+
+    public static Map<MetricID, Metric> getMetricsAsMap(String registryName) throws NoSuchRegistryException, EmptyRegistryException {
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java:68]: Do you expect FFDCs for the thrown exceptions?

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+        MetricRegistry registry = getRegistry(registryName);
+        Map<MetricID, Metric> metricMap = registry.getMetrics();
+        if (metricMap.isEmpty()) {
+            throw new EmptyRegistryException();
+        }
+        return metricMap;
+    }
+
+    public static Map<String, Metadata> getMetricsMetadataAsMap(String registryName) throws NoSuchRegistryException, EmptyRegistryException {
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java:77]: Do you expect FFDCs for the thrown exceptions?

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+        for (String name : registry.getNames()) {
+            metricMetadataMap.put(name, registry.getMetadata(name));
+        }
+        if (metricMetadataMap.isEmpty()) {
+            throw new EmptyRegistryException();
+        }
+        return metricMetadataMap;
+    }
+
+    public static Map<String, Metadata> getMetricsMetadataAsMap(String registryName,
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java:89]: Do you expect FFDCs for the thrown exceptions?

Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java:89]: Exception is caught later and an FFDC ignore is applied there.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java (hunk 26 above)
Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java:77]: Exception is caught later and an FFDC ignore is applied there.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java (hunk 25 above)
Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java:68]: Exception is caught later and an FFDC ignore is applied there.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java (hunk 24 above)
Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/Util30.java:40]: Exception is caught later and an FFDC ignore is applied there.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/BaseMetrics.java (hunk 23 above)
Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/BaseMetrics.java:209]: fixed

**Comment by Channyboy** [comment]
#build
#run-libby-bot

File: dev/com.ibm.ws.microprofile.metrics.3.0/resources/com/ibm/ws/microprofile/metrics/resources/Metrics.nlsprops [hunk 28]
```diff
@@ -0,0 +1,140 @@
+#CMVCPATHNAME N/A
+#COMPONENTPREFIX CWMMC
+#COMPONENTNAMEFOR CWMMC MicroProfile Metrics API 
+#NLS_ENCODING=UNICODE
+#NLS_MESSAGEFORMAT_NONE
+#ISMESSAGEFILE true
+# #########################################################################
+###############################################################################
+# Copyright (c) 2009, 20120 IBM Corporation and others.
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/resources/com/ibm/ws/microprofile/metrics/resources/Metrics.nlsprops:9]: Fix the year

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java [hunk 29]
```diff
@@ -0,0 +1,1090 @@
...
+        }
+        list.add(metricID);
+    }
+
+    public void unRegisterApplicationMetrics() {
+        unRegisterApplicationMetrics(getApplicationName());
+    }
+
+    public void unRegisterApplicationMetrics(String appName) {
+        ConcurrentLinkedQueue<MetricID> list = applicationMap.remove(appName);
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java:253]: You might have a NPE here as `appName` could be `null`.

File: dev/com.ibm.ws.microprofile.metrics.cdi.3.0/bnd.bnd [hunk 30]
```diff
@@ -0,0 +1,68 @@
...
+
+-dsannotations-inherit: true
+-dsannotations: \
+	io.astefanutti.metrics.cdi30.MetricsExtension
+
+
+-buildpath: \
+	com.ibm.ws.logging;version=latest,\
+	com.ibm.ws.adaptable.module;version=latest,\
+	com.ibm.websphere.javaee.cdi.1.2;version=latest,\
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.cdi.3.0/bnd.bnd]: Should cdi be 2.0?

File: dev/com.ibm.ws.microprofile.metrics.cdi.3.0/bnd.bnd [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+	com.ibm.wsspi.org.osgi.service.component.annotations;version=latest,\
+	com.ibm.ws.cdi.interfaces;version=latest,\
+	com.ibm.ws.container.service;version=latest,\
+	com.ibm.ws.container.service.compat;version=latest,\
+	com.ibm.websphere.org.eclipse.microprofile.config.1.3;version=latest,\
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.cdi.3.0/bnd.bnd]: 1.4?

**Comment by Channyboy** [comment]
#build
#run-libby-bot

**Comment by fmhwong** [comment]
TCK is not included in this PR.

**Comment by Channyboy** [comment]
@fmhwong  
The FAT TCKs are avoided as they require additional PRs to complete the feature. The feature as a whole in this PR is `noship`

**Comment by Channyboy** [comment]
The failures do not relate to this PR as this code is "disabled". The failures are pre-existing failures that exist in other Personal Builds and the Continuous build.

File: dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java (hunk 29 above)
Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java:267]: Fixed in https://github.com/OpenLiberty/open-liberty/pull/12299/commits/9edbbeb2b39fae1a4d57c5f9a9c65443ee9ffbc9

File: dev/com.ibm.ws.microprofile.metrics.3.0/resources/com/ibm/ws/microprofile/metrics/resources/Metrics.nlsprops (hunk 28 above)
Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.3.0/resources/com/ibm/ws/microprofile/metrics/resources/Metrics.nlsprops]: Fixed in https://github.com/OpenLiberty/open-liberty/pull/12299/commits/9edbbeb2b39fae1a4d57c5f9a9c65443ee9ffbc9

File: dev/com.ibm.ws.microprofile.metrics.cdi.3.0/bnd.bnd (hunk 31 above)
Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.cdi.3.0/bnd.bnd]: Fixed in the earlier https://github.com/OpenLiberty/open-liberty/pull/12299/commits/6b39b62bde8dcd27d91dad471d48ed340aca470f

File: dev/com.ibm.ws.microprofile.metrics.cdi.3.0/bnd.bnd (hunk 30 above)
Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.cdi.3.0/bnd.bnd]: Fixed in the earlier https://github.com/OpenLiberty/open-liberty/pull/12299/commits/6b39b62bde8dcd27d91dad471d48ed340aca470f

**Comment by Channyboy** [comment]
Squashed review commits, building again but may encounter issues due to current build Conditions (i.e 1)
#build

**Comment by Channyboy** [comment]
Build problems as expected (failed in~1hr)
#build
#run-libby-bot

**Pull request merged**